rls/src/main/java/io/grpc/rls/internal/CachingRlsLbClient.java

```diff
@@ -0,0 +1,962 @@
+ */
+@ThreadSafe
+public final class CachingRlsLbClient {
+
+  private static final Converter<RouteLookupRequest, io.grpc.lookup.v1.RouteLookupRequest>
+      REQUEST_CONVERTER = new RlsProtoConverters.RouteLookupRequestConverter().reverse();
+  private static final Converter<RouteLookupResponse, io.grpc.lookup.v1.RouteLookupResponse>
+      RESPONSE_CONVERTER = new RouteLookupResponseConverter().reverse();
+
+  // All cache status changes (pending, backoff, success) must be under this lock
+  private final Object lock = new Object();
+  // LRU cache based on access order (BACKOFF and actual data will be here)
+  @GuardedBy("lock")
+  private final LinkedHashLruCache<RouteLookupRequest, CacheEntry> linkedHashLruCache;
+  // any RPC on the fly will cached in this map
+  @GuardedBy("lock")
+  private final Map<RouteLookupRequest, PendingCacheEntry> pendingCallCache = new HashMap<>();
+
+  private final SynchronizationContext synchronizationContext;
+  private final ScheduledExecutorService scheduledExecutorService;
+  private final TimeProvider timeProvider;
+  private final Throttler throttler;
+
+  private final LbPolicyConfiguration lbPolicyConfig;
+  private final BackoffPolicy.Provider backoffProvider;
+  private final long maxAgeNanos;
+  private final long staleAgeNanos;
+  private final long callTimeoutNanos;
+
+  private final Helper helper;
+  private final ManagedChannel rlsChannel;
+  private final RouteLookupServiceStub rlsStub;
+  private final RlsPicker rlsPicker;
+  private final ResolvedAddressFactory childLbResolvedAddressFactory;
+  private final RefCountedChildPolicyWrapperFactory refCountedChildPolicyWrapperFactory;
+
+  private CachingRlsLbClient(Builder builder) {
+    helper = checkNotNull(builder.helper, "helper");
+    scheduledExecutorService = helper.getScheduledExecutorService();
+    synchronizationContext = helper.getSynchronizationContext();
+    lbPolicyConfig = checkNotNull(builder.lbPolicyConfig, "lbPolicyConfig");
+    RouteLookupConfig rlsConfig = lbPolicyConfig.getRouteLookupConfig();
+    maxAgeNanos = TimeUnit.MILLISECONDS.toNanos(rlsConfig.getMaxAgeInMillis());
+    staleAgeNanos = TimeUnit.MILLISECONDS.toNanos(rlsConfig.getStaleAgeInMillis());
+    callTimeoutNanos = TimeUnit.MILLISECONDS.toNanos(rlsConfig.getLookupServiceTimeoutInMillis());
+    timeProvider = checkNotNull(builder.timeProvider, "timeProvider");
+    throttler = checkNotNull(builder.throttler, "throttler");
+    linkedHashLruCache =
+        new RlsAsyncLruCache(
+            rlsConfig.getCacheSizeBytes(),
+            builder.evictionListener,
+            scheduledExecutorService,
+            timeProvider);
+    RlsRequestFactory requestFactory = new RlsRequestFactory(lbPolicyConfig.getRouteLookupConfig());
+    rlsPicker = new RlsPicker(requestFactory);
+    rlsChannel = helper.createResolvingOobChannel(rlsConfig.getLookupService());
+    helper.updateBalancingState(ConnectivityState.CONNECTING, rlsPicker);
```

Comment:
Is this run in SynchronizationContext?

Reply:
yes this is created in handleResolvedAddresses.